icechunk/src/session.rs

```diff
@@ -0,0 +1,3288 @@
+        )
+        .await?;
+        ds1.set_chunk_ref(
+            new_array_path.clone(),
+            ChunkIndices(vec![1]),
+            Some(ChunkPayload::Inline("hello1".into())),
+        )
+        .await?;
+
+        let new_array_2_path: Path = "/array_2".try_into().unwrap();
+        ds1.add_array(new_array_2_path.clone(), zarr_meta.clone()).await?;
+        ds1.set_chunk_ref(
+            new_array_2_path.clone(),
+            ChunkIndices(vec![0]),
+            Some(ChunkPayload::Inline("bye0".into())),
+        )
+        .await?;
+
+        let conflicting_snap = ds1.commit("write two chunks with repo 1", None).await?;
+
+        // let's try to create a new commit, that conflicts with the previous one but writes to
+        // different chunks
+        ds2.set_chunk_ref(
+            new_array_path.clone(),
+            ChunkIndices(vec![2]),
+            Some(ChunkPayload::Inline("hello2".into())),
+        )
+        .await?;
+        if let Err(SessionError::Conflict { .. }) =
+            ds2.commit("write one chunk with repo2", None).await
+        {
+            let solver = BasicConflictSolver::default();
+            // different chunks were written so this should fast forward
+            ds2.rebase(&solver).await?;
+            let snapshot = ds2.commit("after conflict", None).await?;
+            let data = ds2.get_chunk_ref(&new_array_path, &ChunkIndices(vec![2])).await?;
+            assert_eq!(data, Some(ChunkPayload::Inline("hello2".into())));
+
+            // other chunks written by the conflicting commit are still there
+            let data = ds2.get_chunk_ref(&new_array_path, &ChunkIndices(vec![0])).await?;
+            assert_eq!(data, Some(ChunkPayload::Inline("hello0".into())));
+            let data = ds2.get_chunk_ref(&new_array_path, &ChunkIndices(vec![1])).await?;
+            assert_eq!(data, Some(ChunkPayload::Inline("hello1".into())));
+
+            // new arrays written by the conflicting commit are still there
+            let data =
+                ds2.get_chunk_ref(&new_array_2_path, &ChunkIndices(vec![0])).await?;
+            assert_eq!(data, Some(ChunkPayload::Inline("bye0".into())));
+
+            let commits = repo.ancestry(&snapshot).await?.try_collect::<Vec<_>>().await?;
+            assert_eq!(commits[0].message, "after conflict");
+            assert_eq!(commits[1].message, "write two chunks with repo 1");
+        } else {
+            panic!("Bad test, it should conflict")
+        }
+
+        // reset the branch to what repo1 wrote
+        let current_snap =
+            fetch_branch_tip(repo.storage().as_ref(), "main").await?.snapshot;
+        update_branch(
+            repo.storage().as_ref(),
+            "main",
+            conflicting_snap.clone(),
+            Some(&current_snap),
+            false,
+        )
+        .await?;
+
+        // TODO: We can't create writeable sessions from arbitrary snapshots anymore so not sure what to do about this?
```

Comment:
it sucks a bit that we cannot, we should try to bring it back after merging

Reply:
Its not a technical limitation, this was the design you and ryan proposed IIRC. Is there a pattern where we would create a writeable session detached from a branch?

Reply:
Ryan doesn't like allowing "anonymous" writes. I do like them and think they are going to be useful. I feel Icechunk as a storage format should not dictate the user's workflow. But we can bring them back later.